ledger/complete/wal/checkpoint_v6_leaf_reader.go

```diff
@@ -0,0 +1,109 @@
+package wal
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/ledger"
+	"github.com/onflow/flow-go/ledger/common/hash"
+	"github.com/onflow/flow-go/ledger/complete/mtrie/flattener"
+	"github.com/onflow/flow-go/ledger/complete/mtrie/node"
+)
+
+type LeafNode struct {
+	Hash    hash.Hash
+	Path    ledger.Path
+	Payload *ledger.Payload
+}
+
+type LeafNodeResult struct {
+	LeafNode *LeafNode
+	Err      error
+}
+
+func nodeToLeaf(leaf *node.Node) *LeafNode {
+	return &LeafNode{
+		Hash:    leaf.Hash(),
+		Path:    *leaf.Path(),
+		Payload: leaf.Payload(),
+	}
+}
+
+func OpenAndReadLeafNodesFromCheckpointV6(dir string, fileName string, logger *zerolog.Logger) (
+	allLeafNodesCh <-chan LeafNodeResult, errToReturn error) {
+
+	filepath := filePathCheckpointHeader(dir, fileName)
+
+	f, err := os.Open(filepath)
+	if err != nil {
+		return nil, fmt.Errorf("could not open file %v: %w", filepath, err)
+	}
+	defer func(file *os.File) {
+		errToReturn = closeAndMergeError(file, errToReturn)
+	}(f)
+
+	subtrieChecksums, _, err := readCheckpointHeader(filepath, logger)
+	if err != nil {
+		return nil, fmt.Errorf("could not read header: %w", err)
+	}
+
+	// ensure all checkpoint part file exists, might return os.ErrNotExist error
+	// if a file is missing
+	err = allPartFileExist(dir, fileName, len(subtrieChecksums))
+	if err != nil {
+		return nil, fmt.Errorf("fail to check all checkpoint part file exist: %w", err)
+	}
+
+	bufSize := 1000
+	leafNodesCh := make(chan LeafNodeResult, bufSize)
+	allLeafNodesCh = leafNodesCh
+	defer func() {
+		close(leafNodesCh)
+	}()
```

Comment:
@koko1123 This got merged but I have a quick question.

What happens when there are more than 1000 (`bufSize`) leaf nodes?  

It seems `allLeafNodesCh` channel's buffer would be full after 1000 nodes, so `processCheckpointSubTrie()` won't be able to send any more `LeafNodeResult` to channel.  And there is no consumer to channel `allLeafNodesCh` because `OpenAndReadLeafNodesFromCheckpointV6` isn't returned.

Reply:
yeah, this is intended. 

We use 1000 leaf nodes, so that the checkpoint reader will be blocked from reading more leaf nodes until the consumer has enough workers (fast enough) to process the remaining leaf nodes in the buffer of the channel. Why? it is so that, if the consumer is slow processing the leaf nodes, such as saving them to a database, we would not read too many leaf nodes into the memory, which otherwise will cause OOM. 

This is mostly for the scenario that an archive node running on commercial hardware being able to process all leaf nodes from the checkpoint file without hitting OOM.

Reply:
@zhangchiqing Thanks for the context.  However, it seems to me that  `OpenAndReadLeafNodesFromCheckpointV6()`  is the creator and the sender for the channel, and there's no consumer of that channel until this function returns.

Given this, if the function gets stuck with full channel then there's no consumer to drain the channel.

One way to resolve this is to make `OpenAndReadLeafNodesFromCheckpointV6()` receive the channel instead of both creating & returning the channel.

Reply:
Good point. Let me fix it in a separate PR

Reply:
Fixed here
https://github.com/onflow/flow-go/pull/4137